storage/remote/queue_manager.go

```diff
@@ -68,9 +73,9 @@ type queueManagerMetrics struct {
 	retriedExemplarsTotal  prometheus.Counter
 	retriedHistogramsTotal prometheus.Counter
 	retriedMetadataTotal   prometheus.Counter
-	droppedSamplesTotal    prometheus.Counter
-	droppedExemplarsTotal  prometheus.Counter
-	droppedHistogramsTotal prometheus.Counter
+	droppedSamplesTotal    *prometheus.CounterVec
+	droppedExemplarsTotal  *prometheus.CounterVec
+	droppedHistogramsTotal *prometheus.CounterVec
```

Comment:
FYI @cstyan the use of a `*prometheus.CounterVec` here makes for _a lot_ of difference in allocations.

> For the following benchmarks, I'm using [this](https://github.com/grafana/prometheus/commit/72e1bba5b86976d6b43c344ccc2b906ca697da39) commit from [this](https://github.com/grafana/prometheus/commits/drop-old-inmemory-samples-benchmark-1/) branch.

To showcase, here's the difference of running BenchmarkSampleSend on the current 7dee9d009 PR HEAD, and versus the same HEAD but with a simple `prometheus.Counter` instead.

```
$ benchstat head-7dee9d009-no-counter-vec.txt head-7dee9d009-counter-vec.txt
...
             │ head-7dee9d009-no-counter-vec.txt │      head-7dee9d009-counter-vec.txt      │
             │             allocs/op             │   allocs/op    vs base                   │
SampleSend-8                          12.00 ± 8%   10030.00 ± 0%  +83483.33% (p=0.000 n=20)
```

There [is already a related fix](https://github.com/prometheus/client_golang/commit/51714a5a30c011624485bd78137378ff0d635383) out on client_golang for this, but will be released on the next v1.18.0 version and it's not currently being used by Prometheus.

When I updated client_golang to include the fix, it dramatically decreased allocations, making it on par with using the simple counter and is going a lot faster when it's discarding all samples.

```
$ go get -u github.com/prometheus/client_golang@51714a5a30c011624485bd78137378ff0d635383
$ ...
$ benchstat head-current-client-golang.txt head-new-client-golang.txt
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/storage/remote
             │ head-current-client-golang.txt │     head-new-client-golang.txt      │
             │             sec/op             │   sec/op     vs base                │
SampleSend-8                      911.1µ ± 1%   735.2µ ± 1%  -19.31% (p=0.000 n=20)

             │ head-current-client-golang.txt │      head-new-client-golang.txt      │
             │              B/op              │     B/op      vs base                │
SampleSend-8                   164.173Ki ± 0%   6.366Ki ± 2%  -96.12% (p=0.000 n=20)

             │ head-current-client-golang.txt │     head-new-client-golang.txt     │
             │           allocs/op            │ allocs/op   vs base                │
SampleSend-8                    10030.00 ± 0%   24.00 ± 4%  -99.76% (p=0.000 n=20)
$ benchstat main-no-filter.txt head-new-client-golang.txt
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/storage/remote
             │ main-no-filter.txt │     head-new-client-golang.txt      │
             │       sec/op       │   sec/op     vs base                │
SampleSend-8        1881.6µ ± 11%   735.2µ ± 1%  -60.93% (p=0.000 n=20)

             │ main-no-filter.txt │      head-new-client-golang.txt      │
             │        B/op        │     B/op      vs base                │
SampleSend-8       250.674Ki ± 8%   6.366Ki ± 2%  -97.46% (p=0.000 n=20)

             │ main-no-filter.txt │     head-new-client-golang.txt     │
             │     allocs/op      │ allocs/op   vs base                │
SampleSend-8          405.00 ± 6%   24.00 ± 4%  -94.07% (p=0.000 n=20)
```

Reply:
On the other hand, I'm happy to see that there's 

a) No performance regression for people _not_ using the flag
b) No performance regression for people using the flag in normal operation, i.e. when all samples go through
c) A significant throughput gain when discarding unwanted samples

Reply:
Given that this is only impacting cases where we're throwing away samples (either via relabeling, due to their age or unintentionally), do you see this as a blocker?

Reply:
@cstyan a new `v1.18.0` release of client_golang is out, which fixes the performance overhead. Do you think it's ok to bump as part of this release, or should we make it a different PR altogether?

https://github.com/prometheus/client_golang/releases/tag/v1.18.0

Reply:
@tpaschalis prometheus should be updating that package soon, there's an automated PR open already for it. Lets just double check benchmark results in your PR after the package update is merged and then we can merge here as well.

Reply:
@cstyan I could only find [this automated PR](https://github.com/prometheus/prometheus/pull/13354
), but it only bumped it in another subpackage and not the main go.mod file.


I've opened the following to manually bump it for the main go.mod and will re-run the benchmarks. https://github.com/prometheus/prometheus/pull/13373

Reply:
client_golang is now updated on main, and I've merged main into this branch to run the benchmarks once more.

Reply:
So after updating client_golang, there's no impact on people _not_ using the feature

```
$ benchstat main.txt head-nofilter.txt
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/storage/remote
             │   main.txt   │       head-nofilter.txt       │
             │    sec/op    │   sec/op     vs base          │
SampleSend-8   2.021m ± 10%   1.999m ± 3%  ~ (p=0.841 n=20)

             │   main.txt   │       head-nofilter.txt        │
             │     B/op     │     B/op      vs base          │
SampleSend-8   276.4Ki ± 6%   279.6Ki ± 6%  ~ (p=0.862 n=20)

             │  main.txt  │      head-nofilter.txt       │
             │ allocs/op  │ allocs/op   vs base          │
SampleSend-8   438.5 ± 4%   440.5 ± 6%  ~ (p=0.507 n=20)
```

as well as a speedup when samples are thrown away
```
$ benchstat head-nofilter.txt head-filter.txt
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/storage/remote
             │ head-nofilter.txt │           head-filter.txt           │
             │      sec/op       │   sec/op     vs base                │
SampleSend-8        1999.5µ ± 3%   723.4µ ± 5%  -63.82% (p=0.000 n=20)

             │ head-nofilter.txt │           head-filter.txt            │
             │       B/op        │     B/op      vs base                │
SampleSend-8      279.571Ki ± 6%   6.304Ki ± 3%  -97.75% (p=0.000 n=20)

             │ head-nofilter.txt │          head-filter.txt           │
             │     allocs/op     │ allocs/op   vs base                │
SampleSend-8         440.50 ± 6%   24.00 ± 4%  -94.55% (p=0.000 n=20)
```